File: plugins/openclaw/src/circuit-breaker.ts
import type { CircuitBreakerState } from "./types.js";

export type CircuitBreakerOptions = {
  failureThreshold: number;
  recoveryIntervalMs: number;
};

/**
 * Simple three-state circuit breaker for the AgentShield HTTP client.
 *
 * State machine:
 *   closed  --(N consecutive failures)--> open
 *   open    --(recoveryIntervalMs elapsed)--> half-open
 *   half-open --(success)--> closed
 *   half-open --(failure)--> open
 */
export class CircuitBreaker {
  private state: CircuitBreakerState = "closed";
  private consecutiveFailures = 0;
  private lastFailureTime = 0;
  private readonly failureThreshold: number;
  private readonly recoveryIntervalMs: number;

  constructor(opts: CircuitBreakerOptions) {
    this.failureThreshold = opts.failureThreshold;
    this.recoveryIntervalMs = opts.recoveryIntervalMs;
  }

  /** Returns true if the circuit is open and calls should be skipped. */
  isOpen(): boolean {
    if (this.state === "closed") {
      return false;
    }

    if (this.state === "open") {
      const elapsed = Date.now() - this.lastFailureTime;
      if (elapsed >= this.recoveryIntervalMs) {
        this.state = "half-open";
        return false; // Allow a probe request
      }
      return true;
    }

    // half-open: allow the probe request through
    return false;
  }

  recordSuccess(): void {
    this.consecutiveFailures = 0;
    this.state = "closed";
  }

  recordFailure(): void {
    this.consecutiveFailures += 1;
    this.lastFailureTime = Date.now();

    if (this.state === "half-open") {
      this.state = "open";
      return;
    }

    if (this.consecutiveFailures >= this.failureThreshold) {
      this.state = "open";
    }
  }

  getState(): CircuitBreakerState {
    // Re-evaluate open -> half-open transition on read
    if (this.state === "open") {
      const elapsed = Date.now() - this.lastFailureTime;
      if (elapsed >= this.recoveryIntervalMs) {
        this.state = "half-open";
      }
    }
    return this.state;
  }
}

File: plugins/hermes/circuit_breaker.py
"""Three-state circuit breaker for the AgentShield HTTP client.

State machine::

    closed  --(N consecutive failures)--> open
    open    --(recovery_interval_ms elapsed)--> half-open
    half-open --(success)--> closed
    half-open --(failure)--> open
"""

from __future__ import annotations

import time
from typing import Literal

CircuitBreakerState = Literal["closed", "open", "half-open"]


class CircuitBreaker:
    """Protects the agent loop from a slow or unavailable engine."""

    def __init__(
        self,
        failure_threshold: int = 3,
        recovery_interval_ms: int = 30_000,
    ) -> None:
        self._state: CircuitBreakerState = "closed"
        self._consecutive_failures: int = 0
        self._last_failure_time: float = 0.0
        self._failure_threshold = failure_threshold
        self._recovery_interval_s = recovery_interval_ms / 1000.0

    # -- public API --------------------------------------------------------

    def is_open(self) -> bool:
        """Return *True* if calls should be skipped (circuit is open)."""
        if self._state == "closed":
            return False

        if self._state == "open":
            elapsed = time.monotonic() - self._last_failure_time
            if elapsed >= self._recovery_interval_s:
                self._state = "half-open"
                return False  # allow a probe request
            return True

        # half-open: allow the probe request through
        return False

    def record_success(self) -> None:
        self._consecutive_failures = 0
        self._state = "closed"

    def record_failure(self) -> None:
        self._consecutive_failures += 1
        self._last_failure_time = time.monotonic()

        if self._state == "half-open":
            self._state = "open"
            return

        if self._consecutive_failures >= self._failure_threshold:
            self._state = "open"

    def get_state(self) -> CircuitBreakerState:
        # Re-evaluate open -> half-open transition on read.
        if self._state == "open":
            elapsed = time.monotonic() - self._last_failure_time
            if elapsed >= self._recovery_interval_s:
                self._state = "half-open"
        return self._state
